### DISCUSSION TITLE
Scheduled Transfers & Transfer Processors

### DISCUSSION THREAD

**Discussion by Kukks** [btcpayserver/btcpayserver#3413] [Ideas / Feature Requests]
This is a feature I am working on for version 1.5.

# Scheduled Transfers 
Scheduled transfers are transactions via a payment method that have been requested to be made by a store in a manner which does not guarantee to be instantaneous. 

When creating a transaction in the on chain wallet, you will have the option to choose "Schedule" instead of "Send".

Pull Payment Payouts, once approved, will now create a scheduled transfer instead of taking the user to the wallet/lightning page to initiate a payment. This plays very well with #1851, which means that with the right transfer processors activated, pull payments payouts can now be automatically paid out. Additionally, this means that makes stores that require its employees to handle and approve refunds, without giving access to the store wallet's funds, possible (cc @lucidddreams).

One can also use this feature to periodically drain a hot wallet in BTCPay Server to another wallet. Simply schedule transfers of an amount that you'd like to drain at, and once your wallet reaches that amount, the transaction will be sent.

With the new CPFP (#3395 ) feature coming in 1.5, we can also incorporate it into Scheduled transfers, so that CPFP gets scheduled. 

# Transfer Processors

Transfer processors are different workflows on how and when scheduled transfers are processed. Each payment type can have multiple different transfer processors.

On chain transfer processors: 
- Automated Processor - this processor runs at an interval and will process as many scheduled transfers as the current wallet balance allows. 
- Payjoin Processor - this processor runs when a payjoin is being proposed by a customer/sender. We batch scheduled transfers directly into the payjoin transaction. The customer's transaction is suddenly a multi-party coinjoin where neither the customer nor the recipients of the scheduled transfers knows what's for who. **Planned for 1.6**
- Manual Processor - When creating a transaction in the on chain wallet, and opting to send straight away, you can choose an option to "batch scheduled transfers into this transaction"

Lightning Transfer processors:
- Automated Processor - this processor runs at an interval and will process as many scheduled transfers as the current wallet balance allows. 
- Manual Processor - A page with a list of all scheduled transfers will appear and you can pay them in one interaction.

**Comment by Kukks** [comment]
cc @Kixunil perhaps we can also use this as a way to do something similar to your [Loptos project](https://github.com/Kixunil/loptos/)

**Reply by Kixunil** [reply]
Wow! Just few minutes ago I had a call with @prusnak where we discussed something like this (among other things) and I discussed this with @MaxHillebrand long time ago as well and more seriously recently (let's implement it reasonably soon serious).

To improve security I'd like to propose this to be a separate daemon, not baked-into BTCPay and not a plugin (will open another discussion about some things soon; edit: done #3414). This is a well-known security practice and will also make it possible to use for non-btcpay users. We were planning to write this in Rust and I still think it's a great choice for security-critical software but if there are some important factors I don't see I'm open to re-evaluation. I will ping some folks interested in this and we will see where things go.

**Reply by Kukks** [reply]
>To improve security I'd like to propose this to be a separate daemon

Sorry but this is a core btcpay feature that will be baked in. The processors could potentially be available through plugins and a plugin that allows remote execution (eg your proposed separate daemon) would do the heavy lifting in a separate manner.
